File: april_challenge/week 4/LRUcache.java
class LRUCache {
    class Node {
        Node prev;
        Node next;
        int key;
        int val;
    }
    // we are using hashmap for fast acess and doubly linked list for fast removal of node
    final Node head = new Node();   // final is bcz we are never going to change it
    final Node tail = new Node();
    Map<Integer, Node> map;
    int cache_capacity;
    
    public LRUCache(int capacity) {
        map = new HashMap(capacity);
        this.cache_capacity = capacity;
        head.next = tail;
        tail.prev = head;
    }
    
    public int get(int key) {
        int result = -1;
        Node node = map.get(key);
        if (node != null) {
            result = node.val;
            remove(node);
            add(node);
        }
        return result;
    }
    
    public void put(int key, int value) {
        Node node = map.get(key);
        if(node != null) {
            remove(node);
            node.val = value;
            add(node);
        } else {
            if (map.size() == cache_capacity) {
                map.remove(tail.prev.key);// remove from hashmap
                remove(tail.prev);// remove from doubly linked list
            }
            Node new_node = new Node();
            new_node.key = key;
            new_node.val = value;
            
            map.put(key, new_node);
            add(new_node);
        }
    }
    
    public void add(Node node) {
        Node head_next = head.next;
        head.next = node;
        node.prev = head;
        node.next = head_next;
        head_next.prev = node;
    }
    
    public void remove(Node node) {
        Node next_node = node.next;
        Node prev_node = node.prev;
        
        next_node.prev = prev_node;
        prev_node.next = next_node;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: april_challenge/week 4/LRUcache.cpp
#include <iostream>
#include <unordered_map>
using namespace std;
class Node
{
public:
    Node *next, *previous;
    int key, value;

    Node()
    {
        next = previous = nullptr;
        key = value = -1;
    }

    Node(int key, int value)
    {
        this->key = key;
        this->value = value;
        next = previous = nullptr;
    }
};

class LRUCache
{
private:
    Node *head, *tail;
    int maxCapacity, currentCapacity;
    unordered_map<int, Node *> uMap;

public:
    LRUCache(int capacity)
    {
        maxCapacity = capacity;
        currentCapacity = 0;
        head = new Node();
        tail = new Node();

        head->previous = tail;
        tail->next = head;
    }

    int get(int key)
    {
        if (uMap.find(key) == uMap.end())
            return -1;
        Node *node = uMap[key];
        int value = node->value;

        uMap.erase(key);
        eraseNode(node);
        addNode(key, value);

        return value;
    }

    void put(int key, int value)
    {
        if (uMap.find(key) != uMap.end())
        {
            currentCapacity--;
            Node *currentNode = uMap[key];
            uMap.erase(key);
            eraseNode(currentNode);
        }
        if (currentCapacity == maxCapacity)
        {
            uMap.erase(tail->next->key);
            eraseNode(tail->next);
            currentCapacity--;
        }
        addNode(key, value);
        currentCapacity++;
    }

    void eraseNode(Node *node)
    {
        node->previous->next = node->next;
        node->next->previous = node->previous;
        delete node;
    }

    void addNode(int key, int value)
    {
        Node *node = new Node(key, value);
        node->next = head;
        node->previous = head->previous;
        head->previous->next = node;
        head->previous = node;
        uMap[key] = node;
    }
};